core/drivers/riscv_zkr_rng.c

```diff
@@ -4,46 +4,64 @@
  */
 
 #include <crypto/crypto.h>
+#include <encoding.h>
+#include <kernel/delay.h>
 #include <kernel/panic.h>
 #include <riscv.h>
 #include <rng_support.h>
 #include <tee/tee_cryp_utl.h>
 
-#define OPST_BIST 0b00
-#define OPST_WAIT 0b01
-#define OPST_ES16 0b10
-#define OPST_DEAD 0b11
+#define RNG_TIMEOUT_US	1000000
 
-TEE_Result hw_get_random_bytes(void *buf, size_t len)
+static inline bool __must_check seed_get_random_u16(uint16_t *val)
 {
-	uint8_t *ptr = buf;
-	uint32_t val = 0;
+	uint64_t timeout = timeout_init_us(RNG_TIMEOUT_US);
+	uint32_t seed = 0;
+	uint32_t opst = 0;
 
-	while (len > 0) {
+	do {
 		/*
-		 * The seed register must be accessed using CSR read-write
-		 * instructions. The write operation is ignored and serves
-		 * to indicate polling and flushing.
+		 * The seed register must be accessed using CSR
+		 * read-write instructions. The write operation
+		 * is ignored and serves to indicate polling and
+		 * flushing.
 		 */
-		val = swap_csr(CSR_SEED, val);
+		seed = swap_csr(CSR_SEED, 0);
+		opst = seed & SEED_OPST;
 
-		switch (val >> 30) {
-		case OPST_BIST:
-		case OPST_WAIT:
-			continue;
-		case OPST_ES16:
-			*ptr++ = val & 0xff;
-			len--;
-			if (len > 0) {
-				*ptr++ = val >> 8;
-				len--;
-			}
-			break;
-		case OPST_DEAD:
+		switch (opst) {
+		case SEED_OPST_ES16:
+			*val = seed & SEED_ENTROPY;
+			return true;
+		case SEED_OPST_DEAD:
 			/* Unrecoverable self-test error */
-			return TEE_ERROR_BAD_STATE;
+			return false;
+		case SEED_OPST_BIST:
+		case SEED_OPST_WAIT:
 		default:
-			break; /* can't happen */
+			cpu_relax();
+			continue;
+		}
+	} while (!timeout_elapsed(timeout));
```

Comment:
In a thread context, this timeout can be triggered due to thread rescheduling.

Reply:
He's right, maybe the best thing would be to put after the loop:
`
seed = swap_csr(CSR_SEED, 0);
opst = seed & SEED_OPST;
if (opst != SEED_OPST_ES16) {
        EMSG("Failed to produce a sufficient amount of entropy");
        return false;
}

*val = seed & SEED_ENTROPY

return true;
`

what do you think?

Reply:
> In a thread context, this timeout can be triggered due to thread rescheduling.

Do you mean the foreign interrupts ?
More detail please.

Reply:
Hi All,

Here is a problem, we are using mtime as generic timer which is shared between REE and TEE. Given 1-second timeout, If foreign interrupts occur, the time available for the while loop will definitely be less than 1 second, right? Even worse, if it returns to TEE after more than 1 second, it will directly timeout and no change to retry the seed.

Using a retry counter still seems necessary? Any thoughts?

Reply:
Yes, this is related to foreign interrupts. A TEE thread can be interrupted by a foreign interrupt. When so, the TEE thread is suspended and CPU goes back to normal world (let's say to the Linux kernel). The TEE thread is resumed when the Linux kernel schedules back the thread related to that TEE thread.
When `timeout_elapsed()` function is called from an interruptible TEE thread context and it returns `true`, caller of that function should be aware that the condition he/she expects may be already reached and that the timeout out was triggered only because the Linux kernel scheduler scheduled some other thread(s) before resuming the thread related to that TEE thread.
For example this a why helper macro `IO_READ32_POOL_TIMEOUT()` was implemented.

I was to give some example based on the current implementation of this P-R, but I see that actually the `while()` instruction at line 45 is never reached. Indeed, the block either exits from `return true` (line 35) / `return false` (line 38) or spins back from `continue` (line 43)` even before we reach line 45. Therefore my suggestion (feel free to implement something else):
```patch
	do {
		/*
		 * The seed register must be accessed using CSR
		 * read-write instructions. The write operation
		 * is ignored and serves to indicate polling and
		 * flushing.
		 */
		seed = swap_csr(CSR_SEED, 0);

		opst = seed & SEED_OPST;

		switch (opst) {
		case SEED_OPST_ES16:
			*val = seed & SEED_ENTROPY;
			return true;
		case SEED_OPST_DEAD:
			/* Unrecoverable self-test error */
			return false;
		case SEED_OPST_BIST:
		case SEED_OPST_WAIT:
		default:
			cpu_relax();
-			continue;
		}
	} while (!timeout_elapsed(timeout))
+
+	/* Consider timeout case due to normal world scheduler */
+	seed = swap_csr(CSR_SEED, 0);
+	if (seed & SEED_OPST == SEED_OPST_ES16) {
+		*val = seed & SEED_ENTROPY;
+		return true;
+	}

	return false;
 }
```

Note: `cpu_relax()` is not implemented in OP-TEE. If there a cpu-pause instruction for RiscV, i'd rather suggest to name the helper fonction something like `riscv_cpu_pause()` to avoid confusion.

Reply:
Yes @lyctw, this is why we suggested to the direct timeout case. It is also the case for STM32 platforms and we proceed the same way.

@etienne-lms 's implementation is looking good to me. Maybe the comment can be rephrased. You can also keep your error message to explicit what failed.

Reply:
Hi @etienne-lms 

> I see that actually the while() instruction at line 45 is never reached

Nice catch! Will fix.

Reply:
```
	} while (!timeout_elapsed(timeout))
+
+	/* Consider timeout case due to normal world scheduler */
+	seed = swap_csr(CSR_SEED, 0);
+	if (seed & SEED_OPST == SEED_OPST_ES16) {
+		*val = seed & SEED_ENTROPY;
+		return true;
+	}
```

this change is equal to:

```
	} while (!timeout_elapsed(timeout))
	
+       retry_count = 1;
+       do {
+           // read seed
+       } while(retry_count--);
```

If the last retry fails, it returns false right away, regardless of the OPST state, which may still allow the next retry.
I am leaning towards a retry counter as I originally implemented, as the seed and timer are shared. We simply cannot use timeout and `cpu_relax()` should've addressed bandwidth issue. 

I hope that makes sense.

Reply:
I still think the timeout is more relevant as long as gathering the entropy is based on time :) If the timeout value expires, you read the seed state and that there's still no entropy available, you should expect to fail.

Based on the description, the RISCV pause instruction you suggested seemed appropriate for a cpu_relax()-like function. It can even be implemented in RISCV arch helpers IMO.